File: vudroid/src/main/java/org/vudroid/core/DecodeTask.java
package org.vudroid.core;

import android.graphics.RectF;

/**
 * @author: archko 2024/9/2 :14:43
 */
public class DecodeTask {
    public static final int TYPE_PAGE = 0;
    public static final int TYPE_NODE = 1;
    public final String decodeKey;
    public final PageTreeNode node;
    public final int pageNumber;
    public final int type;
    public final float zoom;
    public final DecodeService.DecodeCallback decodeCallback;
    public final RectF pageSliceBounds;
    public boolean crop = true;
    public int dpi = 0;
    public final float fontSize;

    public DecodeTask(PageTreeNode node, boolean crop, int pageNumber, DecodeService.DecodeCallback decodeCallback, float zoom, String decodeKey, RectF pageSliceBounds, int dpi, float fontSize) {
        this.node = node;
        this.crop = crop;
        this.type = node == null ? TYPE_PAGE : TYPE_NODE;
        this.pageNumber = pageNumber;
        this.decodeCallback = decodeCallback;
        this.zoom = zoom;
        this.decodeKey = decodeKey;
        this.pageSliceBounds = pageSliceBounds;
        this.dpi = dpi;
        this.fontSize = fontSize;
    }

    @Override
    public String toString() {
        return "Task{" +
                "page=" + pageNumber +
                ", zoom=" + zoom +
                ", bounds=" + pageSliceBounds +
                ", Key=" + decodeKey +
                '}';
    }
}

File: readercommon/src/main/java/cn/archko/pdf/core/component/DecodeTask.kt
package cn.archko.pdf.core.component

import android.graphics.Bitmap
import android.graphics.Rect
import android.graphics.RectF
import cn.archko.pdf.core.entity.APage

/**
 * 解码任务
 * @author: archko 2025/1/10
 */
data class DecodeTask(
    val type: TaskType,
    val pageIndex: Int,
    val key: String,
    val aPage: APage,
    val zoom: Float = 1f,
    val pageSliceBounds: RectF,
    val width: Int,
    val height: Int,
    val crop: Boolean = false,
    val callback: DecodeCallback? = null,
) {
    override fun toString(): String {
        return "DecodeTask(type=$type, pageIndex=$pageIndex, key='$key', zoom=$zoom, pageSliceBounds=$pageSliceBounds, width=$width, height=$height, crop=$crop, aPage=$aPage)"
    }
}

enum class TaskType {
    PAGE,
    NODE,
    CROP
}

/**
 * 解码回调接口
 */
interface DecodeCallback {
    fun onDecodeComplete(bitmap: Bitmap?, isThumb: Boolean, error: Throwable?)
    fun shouldRender(pageNumber: Int, isFullPage: Boolean): Boolean
    fun onFinish(pageNumber: Int)
}

/**
 * 解码器接口
 */
interface Decoder {
    suspend fun decodePage(task: DecodeTask): Bitmap?
    suspend fun decodeNode(task: DecodeTask): Bitmap?
    suspend fun processCrop(task: DecodeTask): CropResult?
    fun generateCropTasks(): List<DecodeTask>
}

/**
 * 切边处理结果
 */
data class CropResult(
    val pageIndex: Int,
    val cropBounds: RectF?
)
